src/com/intellij/idea/plugin/hybris/properties/impl/PropertiesServiceImpl.kt

```diff
@@ -142,52 +104,114 @@ class PropertiesServiceImpl(val project: Project) : PropertiesService {
                     "env.properties" -> envPropsFile = file
                     "advanced.properties" -> advancedPropsFile = file
                     "local.properties" -> localPropsFile = file
-                    else -> {
-                        for (property in file.properties) {
-                            if (property.key != null) {
-                                result[property.key!!] = property
-                            }
-                        }
-                    }
+                    "project.properties" -> propertieFiles.add(file)
                 }
             }
-        addPropertyFile(result, envPropsFile)
-        addPropertyFile(result, advancedPropsFile)
-        addPropertyFile(result, localPropsFile)
 
-        val optDir = result[HybrisConstants.PROPERTY_OPTIONAL_CONFIG_DIR]
-        addOptionalConfiguration(result, optDir)
+        envPropsFile?.let { propertieFiles.add(0, it) }
+        advancedPropsFile?.let { propertieFiles.add(1, it) }
+        localPropsFile?.let { propertieFiles.add(it) }
+
+        propertieFiles.forEach { file -> addPropertyFile(result, file) }
+
+        loadHybrisRuntimeProperties(result)
+        loadHybrisOptionalConfigDir(result)
 
         return ArrayList(result.values)
     }
 
+    private fun findAllProperties(): LinkedHashMap<String, String> {
+
+        val result = LinkedHashMap<String, String>()
+        findAllIProperties().forEach { property ->
+            property.value?.let {
+                result[property.key!!] = it
+            }
+        }
+
+        replacePlaceholders(result)
 
+        return result
+    }
 
-    private fun addOptionalConfiguration(result: LinkedHashMap<String, IProperty>, optDir: IProperty?) {
-        if (optDir == null) {
-            return
+    private fun replacePlaceholders(result: LinkedHashMap<String, String>) {
+        result.forEach { key, value ->
+            if (value.contains(nestedPropertyPrefix)) {
+                replacePlaceholder(result, key, value);
+            }
         }
-        val dir = File(optDir.value)
-        if (!dir.isDirectory) {
-            return
+    }
+
+    private fun replacePlaceholder(result: LinkedHashMap<String, String>, key: String, value: String) {
+        val pattern = Pattern.compile("\\$\\{[^\\}]*\\}")
+        val matcher = pattern.matcher(value)
+        matcher?.let {
+            while (it.find()) {
+                val placeholder: String = it.group()
+                val nestedKey: String = placeholder.substring(nestedPropertyPrefix.length, placeholder.length - nestedPropertySuffix.length)
+                val nestedValue: String? = result[nestedKey]
+                nestedValue?.let {
+                    var newValue = nestedValue
+                    if (nestedValue.contains(nestedPropertyPrefix)) {
+                        replacePlaceholder(result, nestedKey, nestedValue)
```

Comment:
this is risky, dev can declare circular dependency by defining something like:
```
myProp=${nestedProp}
nestedProp=${myProp}
```

Reply:
I will prevent it in the Plugin, but Devs should be aware that SAP does not prevent this.